### PR TITLE
Make `Range` instances somewhat immutable

### PR DESCRIPTION
**Pull request by quicklizard99**
This PR addresses two problems with `Range`.

1. Attributes of `Range` instances can be set, bypassing validation and violating class invariants:

    ```pycon
    >>> from xocto import ranges
    >>> r = ranges.Range(0, 1)
    >>> r.start = 2
    >>> r.boundaries = '😱'
    ```

2. Additional attributes can be set on instances of classes derived from `Range`

    `Range` uses `__slots__`, which prevents creation of instance dictionaries. Classes derived from `Range` do not have `__slots__ = ()`, which means that instance dictionaries are created:

    ```pycon
    >>> from xocto import localtime
    >>> r = ranges.FiniteDatetimeRange(localtime.datetime(2000, 1, 1), localtime.datetime(2000, 1, 3))
    >>> r.some_other_thing = "I shouldn't be here"
    >>> r.__dict__
    {'some_other_thing': "I shouldn't be here"}
    ```

Fix these problems:

* add `__slots__ = ()` to classes that have `Range` as a base class (preventing creation of instance dicts),
* add `Range.__setattr__` to raise `AttributeError`, and
* alter `Range.__init__` to use the same `object.__setattr__` [trick as `attrs`](https://www.attrs.org/en/stable/init.html#post-init) to set attributes during initialisation.

### CODE REVIEW

File: xocto/ranges.py [hunk 1]
```diff
@@ -143,15 +143,23 @@ class Range(Generic[T]):
...
-    ]
+    )
+
+    start: Optional[T]
+    end: Optional[T]
+    boundaries: RangeBoundaries
+    _is_left_exclusive: bool
+    _is_left_inclusive: bool
+    _is_right_exclusive: bool
+    _is_right_inclusive: bool
```
Comment (quicklizard99) [inline xocto/ranges.py:163]: These types were previously implicitly declared when `__init__` set attributes.

File: xocto/ranges.py [hunk 2]
```diff
@@ -473,8 +484,13 @@ class HalfFiniteRange(Range[T]):
     HalfFiniteRange.
     """
 
+    __slots__ = ()
+
+    # Redefine types in base class
     start: T
-    boundaries = RangeBoundaries.INCLUSIVE_EXCLUSIVE
```
Comment (quicklizard99) [inline xocto/ranges.py:477]: Setting the default value using a class attribute does not play with `__slots__`. I have moved this default to `__init__`.

File: xocto/ranges.py [hunk 3]
```diff
@@ -229,6 +236,14 @@ def __str__(self) -> str:
     def __repr__(self) -> str:
         return f"<Range: {str(self)}>"
 
+    def __setattr__(self, name: str, value: Any) -> None:
+        if name in type(self).__slots__:
+            raise AttributeError("Can't set attributes")
+        else:
+            raise AttributeError(
+                f"'{type(self).__name__}' object has no attribute '{name}'"
+            )
```
Comment (quicklizard99) [inline xocto/ranges.py:245]: These exception messages consistent with those raised by `collections.namedtuple`.

File: xocto/ranges.py [hunk 4]
```diff
@@ -221,6 +236,14 @@ def __str__(self) -> str:
...
```
Comment (jamesbeith) [inline xocto/ranges.py:245]: Perhaps explicit is better than implicit, but I wonder if we could delegate this branch to `super()`?

Comment (jamesbeith) [inline xocto/ranges.py:245]: Perhaps explicit is better than implicit, but I wonder if we could delegate this branch to `super()`?

**Review by jarshwah — APPROVED** [review]
Good change, thanks for making it. As this is a potentially breaking change we'll need to at least bump the minor version when releasing.

File: xocto/ranges.py (hunk 4 above)
Comment (quicklizard99) [reply xocto/ranges.py]: That is a good point. Change to use `super`

**Pull request merged**